File: ex449.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

import java.util.*;

public class Codec {
    
    // Get the postorder traverse sequence of the tree
    private String helper(TreeNode root) {
        // Postorder traverse
        if (root == null)    return "";
        
        String data = "";
        String left = helper(root.left), right = helper(root.right);
        if (left != "")    data = data + "," + left;
        if (right != "")    data = data + "," + right;
        data = data + "," + String.valueOf(root.val);
        
        return data;
    }

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        String res = helper(root);
        StringBuilder sb = new StringBuilder(res);
        
        int i = 1;
        while (i < sb.length()) {
            if (sb.charAt(i) == ',' && sb.charAt(i - 1) == ',') {
                sb.deleteCharAt(i);
            } else {
                i++;
            }    
        }
        if (sb.length() != 0 && sb.charAt(0) == ',')    sb.deleteCharAt(0);
        String data = sb.toString();
        System.out.println(data);
        return data;
    }

    // Get the original tree from the array
    private TreeNode helper2(int[] array) {
        if (array.length == 0)    return null;
        
        int root = array[array.length - 1];
        int index = -1;
        for (int i = 0; i < array.length; i++) {
            if (array[i] > root) {
                index = i;
                break;
            }
        }
        if (index == -1)    index = array.length - 1;
        
        TreeNode rootNode = new TreeNode(root);
        TreeNode leftNode = helper2(Arrays.copyOfRange(array, 0, index));
        TreeNode rightNode = helper2(Arrays.copyOfRange(array, index, array.length - 1));
        rootNode.left = leftNode;
        rootNode.right = rightNode;
        
        return rootNode;
    }
    
    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data.length() == 0)    return null;
        
        String[] sarray = data.split(",");
        int[] array = new int[sarray.length];
        for (int i = 0; i < array.length; i++) {
            array[i] = Integer.parseInt(sarray[i]);
        }
        
        return helper2(array);
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));

File: code/ex449.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string s = "";
        dfs1(root, s);
        return s;
    }
    
    void dfs1(TreeNode* root, string& s) {
        if (root == NULL) {
            s += "# ";
            return;
        }
        s += to_string(root->val) + " ";
        dfs1(root->left, s);
        dfs1(root->right, s);
    }

    // Decodes your encoded data to tree.
    int i = 0;
    
    TreeNode* deserialize(string data) {
        return dfs2(data);
    }
    
    TreeNode* dfs2(string data) {
        string cur = getNext(data);
        if (cur == "#")    return NULL;
        TreeNode* root = new TreeNode(stoi(cur));
        root->left = dfs2(data);
        root->right = dfs2(data);
        return root;
    }
    
    string getNext(string s) {
        string res = "";
        while (s[i] != ' ') {
            res += s[i++];
        }
        i++;
        return res;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec* ser = new Codec();
// Codec* deser = new Codec();
// string tree = ser->serialize(root);
// TreeNode* ans = deser->deserialize(tree);
// return ans;